### PR TITLE
fix(records): one minor-unit scale, and a test that keeps it singular

### PR DESCRIPTION
**Pull request by tinnhlt-gra**
## The finding

A currency's minor-unit scale — 100 for EUR, 1000 for KWD, 1 for VND — is one
fact. The **table** holding it was already shared, and its own comment said so:

> one spelling of the ISO fact, never a second copy that could drift

That claim was true about the table and **false about the arithmetic beside
it**. Four functions each built the power of ten themselves:

| | |
|---|---|
| `compose/offerdraft_price.go` | the loop, then `values.MajorUnits` reimplemented verbatim to render the result |
| `modules/overlay/transforms.go` | the loop, feeding `decimalStringToMinor` |
| `overlay/hubspot/mapwrite.go` | a whole **second renderer**, by string splice rather than divmod |
| `shared/kernel/values/minorunits.go` | the one that belongs there |

**They agreed.** Before deleting the fourth I compared it against
`values.MajorUnits` over every combination of 17 currency codes (including the
zero-decimal, three-decimal, four-decimal and unknown ones) and 16 amounts
(including `math.MinInt64` and `math.MaxInt64`) — **272 pairs, 0 disagreed.**

That is exactly how long a second copy stays harmless: until the commit that
corrects one currency's digit count in the table reaches only the callers
somebody remembered.

## The fix

`values` already exported almost everything the other three needed —
`MajorUnits` for the decimal string, `WholeMajorUnits` for the truncated
integer. Added `MinorUnitScale` for the multiplier itself, which is the one
shape a caller converting *into* minor units needs and that the two dividers
cannot supply.

Deleting the fourth copy made `overlay.ISO4217MinorUnitExponent` **dead** — the
wrapper whose doc comment named the HubSpot write mapping as the reason it was
exported, and named *"one spelling of the ISO fact, never a second copy that
could drift"* as its own promise. Its last caller **was** the second renderer.
Removed.

The private helper is `powerOfTen(digits)` rather than `minorUnitScale(digits)`
— two names one capital apart for two different signatures is a reader's
problem before it is `revive`'s.

## The gate is a census of zero, so the detector is tested separately

This is the hard kind to write honestly: a gate asserting a shape is **absent**
passes identically when the shape is absent and when the detector has stopped
detecting. The census alone would be worth nothing.

So `TestTheHandScaledDetectorSeesWhatItClaimsTo` reads the detector directly —
**11 cases, both directions**:

*Must fire* — the loop as it shipped in compose; the loop as it shipped in
overlay with the digits read inline; a `range` loop; the multiplication written
out as `scale = scale * 10`; `math.Pow10`, the next spelling nobody has written
yet; and an unqualified call, as `values` itself would write it.

*Must not* — asking the table without scaling; asking and then delegating the
scaling to `values`; scaling by ten for something that is not money; a
**hundred**, which is `check-money-scale.sh`'s subject and not this one (two
gates claiming one shape is how a finding gets reported twice and fixed
neither time); and the two halves in **different functions** of one file.

The census also refuses a walk that finds fewer than 500 Go files, because a
walk that finds nothing certifies nothing.

## Verified against the real defect, not just a planted string

Reverting the compose call site to its loop — keeping the build green, so the
gate is what fails and not the compiler:

```
--- FAIL: TestNobodyBuildsAMinorUnitScaleByHand
    these functions ask values for a currency's minor-unit digit count and then
    build the power of ten themselves:
      internal/compose/offerdraft_price.go: priceEvidencedInSnippet
```

Restored → green. `make check` green.

## Why per-function and not per-file

A file may hold a renderer that asks the table and a separate helper that
scales something unrelated by ten; asking whether both appear *somewhere* in
the same file reports a pairing nobody wrote. Per-declaration is also the
reading that fails safe — it can miss a defect split across two functions, and
it cannot invent one. That case is planted.

Found while auditing the merged result of #2308 and #2317.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Unifies currency minor‑unit scaling and rendering under `backend/internal/shared/kernel/values` and hardens a test gate to keep a single source of truth. Previously callers built 10^digits or rendered decimals locally; now they call `values.MajorUnits`, `values.WholeMajorUnits`, or `values.MinorUnitScale`. No user‑visible behavior change.

- Replaced local logic in compose price evidence, overlay transforms, and HubSpot map write with `values.MajorUnits`, `values.WholeMajorUnits`, and `values.MinorUnitScale`.
- Removed `overlay.ISO4217MinorUnitExponent` and the local `minorToDecimalString`.
- Added `values.MinorUnitScale`; internal helper renamed to `powerOfTen`, and `MajorUnits`/`WholeMajorUnits` use it.
- Added `backend/minorunitscale_test.go`: an AST gate that flags hand‑built scales (loops or `math.Pow10`) and decimal‑point renderers keyed on a digit‑count parameter; resolves `values`/`math` qualifiers via imports; the renderer arm no longer requires a `values` import; consolidates import resolution into `importAlias`; rejects incomplete walks (<500 Go files); includes positive/negative probes.
- Style: gofumpt formatting on the extracted probe table; no functional changes.

**Migration**
- Replace `overlay.ISO4217MinorUnitExponent` with `values.MinorUnitDigits` or `values.MinorUnitScale`; render with `values.MajorUnits`.

<sup>Written for commit c0ca14e7e01c5a7c69e16eabbf82faa1d751b632. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/margince/margince/pull/2349?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Improvements**
  * Standardized currency amount formatting across pricing details and HubSpot deal amounts.
  * Improved accuracy and consistency for currencies with different decimal precision.
  * Enhanced handling of whole and fractional currency values for clearer amount displays.

* **Quality**
  * Added automated safeguards to help prevent inconsistent currency-scaling and formatting behavior.
  * Strengthened validation to maintain reliable formatting across supported currencies and amount displays.

### CODE REVIEW

**Comment by ancorn-gradion** [comment]
## UAT — the census against each copy it deleted

https://github.com/user-attachments/assets/f7c76c3f-9c85-49e6-9845-456a1db4d736

*A replay of a recorded terminal session, not a live screen capture: the transcript is real output, painted into frames so it is readable at video speed.*

> **Updated after three further review rounds.** The recording is still accurate about what it shows — each deleted copy restored and named — but the detector has been tightened five times since, and the case count below has moved from 15 to **20**. The original text is superseded rather than removed, and the difference is the interesting part: see *What review changed after this was recorded*.

Each of the deleted copies is restored **in the real tree**, the build kept green so that a failure is the **gate** and not the compiler, then removed again. The census names each one:

| restored copy | census |
|---|---|
| the loop in `compose/offerdraft_price.go` | `FAIL` — names `priceEvidencedInSnippet` |
| the loop in `modules/overlay/transforms.go` | `FAIL` — names `transformAmountMinorByCurrency` |
| the string-splice renderer in `overlay/hubspot` | `FAIL` — names `minorToDecimalString` |

**The third one is the point of this UAT.** It builds no power of ten — it splices a decimal point into a formatted integer — so the arithmetic arm was *structurally* blind to it. Review found the census green over exactly the function this PR deletes, and the fix was a second arm keyed on the **parameter**: a renderer is *handed* the digit count rather than asking for it, which is precisely why nobody had spotted the copy either.

That generalises past this PR: **the copies you delete are not all written the same way, and the one that escaped notice longest escaped for a reason your detector probably shares.** Planting each shape is necessary but not sufficient — the plants are written by the same author as the detector and inherit its blind spot. Only restoring them in the real tree opened it.

It also had a blast radius. Two comments claimed `values.MajorUnits` is "the one renderer of a minor-unit integer as a currency-decimal string", and this repo forbids a uniqueness claim no test holds. A blind census does not merely miss a duplicate — it quietly converts every comment resting on it into a false claim.

## What review changed after this was recorded

Three further rounds, and the through-line is that **every finding was the detector naming the wrong subject, not missing one** — which is the worse direction for a census, because a false finding sends somebody to disprove it.

| what it did | what it does now |
|---|---|
| fired on any `* 10` anywhere in a function that read the digit count | the multiplication must be **inside a loop** |
| matched any `x.MinorUnitDigits(…)` by selector name | the **values qualifier** is resolved through the file's imports |
| accepted a bare `MinorUnitDigits` in any package not importing values | the file must **be** the values package |
| matched any `x.Pow10` | must be **math**'s |
| let the digit-count parameter satisfy "reaches the render" from the **signature** | walks `fn.Body` only |

That last one is the one I would flag: `ast.Inspect(fn, …)` includes `Type.Params`, so the parameter's own identifier satisfied the guard before the body was ever read. **A guard that cannot fail** — in the file written to stop exactly that.

The probes gained a third rendering mode for the same reason the second existed: the detector's answers depend on which file the code is in, so a probe that guesses wrong asks a different question than the tree does. `""` is an ordinary caller importing values, `"values"` is the values package itself, `"noimport"` is a package that does not use values at all and may declare its own `MinorUnitDigits`.

## The detector, read directly — now 20 cases

A census of zero passes identically over a clean tree and over a detector that detects nothing, so the detector is read straight, both directions, all green.

*Fires* — the loop as it shipped in compose; the loop as it shipped in overlay with the digits read inline; a `range` loop; the multiplication written out; **the string-splice renderer verbatim**; the same renderer with `Sprintf`; `math.Pow10`; an unqualified call inside values.

*Silent* — asking the table without scaling; asking then delegating to `values`; scaling by ten for something that is not money; a **hundred** (`check-money-scale.sh`'s subject, not this one); a digit count that renders nothing; a digit count the render **ignores**; a renderer whose digit count is **signature-only**; reading the digit count beside an unrelated `* 10`; another package's `MinorUnitDigits`; another package's **own unqualified** one; a decimal point with no digit count in sight; the two halves in **different functions** of one file.

**Mutation-verified: five mutants, one per guard, all killed.** Three of them survived before these cases existed — which is the only reason I know the cases are worth anything.

```
$ go build ./... && go test . -run 'MinorUnitScale|HandScaled'
ok  	github.com/gradionhq/margince/backend
```

Tree left clean — `git status --porcelain` empty after every restoration.

File: backend/internal/shared/kernel/values/minorunits.go [hunk 1]
```diff
@@ -121,19 +121,38 @@ func MajorUnits(amountMinor int64, currency string) string {
...
+// a decimal figure INTO minor units needs, which MajorUnits and WholeMajorUnits
+// cannot supply because they only ever divide by it.
+//
+// It is exported for that one shape and no other. A caller that wants to RENDER
+// an amount wants MajorUnits or WholeMajorUnits instead — reaching for the scale
+// to do the division by hand is how a currency's digit count comes to be right
+// in this table and wrong at the surface reading it.
+func MinorUnitScale(currency string) int64 {
+	return powerOfTen(MinorUnitDigits(currency))
+}
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/internal/shared/kernel/values/minorunits.go:154]: REFUTED. There is no such repository rule. CLAUDE.md's DO NOT TOUCH list is specific and short — the two generated files (`internal/contracts/api_gen.go`, `internal/compose/stubs_gen.go`), shipped core migrations, the `database.WithWorkspaceTx` GUC contract, and `internal/shared/apperrors`. A rule barring edits to every backend Go file would contradict the same document's own contributing loop, which describes branching, editing, running `make check` and opening a PR against exactly this tree, and would make the repository unmaintainable. This finding has appeared before on another PR today and was refuted the same way. Nothing to revert.

File: backend/minorunitscale_test.go [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+// unit at all — and those are not this gate's business. What it may not do is
+// take the digits and do values' job with them, because values exports every
+// shape a caller here has needed: MinorUnitScale for the multiplier,
+// MajorUnits for the decimal string, WholeMajorUnits for the truncated integer.
+//
+// This is a census of ZERO, and that is the hard kind to write honestly. A gate
+// asserting a shape is absent passes identically when the shape is absent and
+// when the detector is broken, so the census below is worth nothing without
+// TestTheHandScaledDetectorSeesWhatItClaimsTo beneath it — which plants all
+// four shapes that actually shipped, plus the lookalikes that must NOT be
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go:31]: REFUTED. There is no such repository rule. CLAUDE.md's DO NOT TOUCH list is specific and short — the two generated files (`internal/contracts/api_gen.go`, `internal/compose/stubs_gen.go`), shipped core migrations, the `database.WithWorkspaceTx` GUC contract, and `internal/shared/apperrors`. A rule barring edits to every backend Go file would contradict the same document's own contributing loop, which describes branching, editing, running `make check` and opening a PR against exactly this tree, and would make the repository unmaintainable. This finding has appeared before on another PR today and was refuted the same way. Nothing to revert.

File: backend/minorunitscale_test.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+}
+
+// buildsAPowerOfTen reports whether the declaration multiplies an accumulator
+// by ten, or reaches for math.Pow10.
+func buildsAPowerOfTen(fn *ast.FuncDecl) bool {
+	builds := false
+	ast.Inspect(fn, func(n ast.Node) bool {
+		switch v := n.(type) {
+		case *ast.AssignStmt:
+			if v.Tok == token.MUL_ASSIGN && anyLiteralTen(v.Rhs) {
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go:296]: Correct, and it is the worse direction for a census — a false finding sends somebody to disprove it, which costs more than a missed one. The multiplication must now sit inside a LOOP. The detector still cannot prove the loop is bounded BY the digit count (that needs data flow, and the test says so rather than implying otherwise), but "in a loop" is what separates building a scale from ordinary arithmetic and it is checkable. Planted as `reading the digit count beside an unrelated * 10` — a finding before, silent now — and all three deleted copies are still named when restored in the real tree.

File: backend/minorunitscale_test.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	asks := false
+	ast.Inspect(fn, func(n ast.Node) bool {
+		call, ok := n.(*ast.CallExpr)
+		if !ok {
+			return true
+		}
+		switch f := call.Fun.(type) {
+		case *ast.Ident:
+			asks = asks || f.Name == digitsFunc
+		case *ast.SelectorExpr:
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go:238]: Correct. The qualifier is resolved through the file's imports now, and a bare call counts only where the values package itself would write one. Planted as `a different package's MinorUnitDigits is not values'`. Worth recording what that exposed: the planted probes did not import values at all, so tightening the match turned every detection case red at once — a probe without the import asks a different question than the tree does. They carry the real imports now, and the one case that must be unqualified says so explicitly.

File: backend/minorunitscale_test.go [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+// by ten, or reaches for math.Pow10.
+func buildsAPowerOfTen(fn *ast.FuncDecl) bool {
+	builds := false
+	ast.Inspect(fn, func(n ast.Node) bool {
+		switch v := n.(type) {
+		case *ast.AssignStmt:
+			if v.Tok == token.MUL_ASSIGN && anyLiteralTen(v.Rhs) {
+				builds = true
+			}
+			// `scale = scale * 10` written out, and the `:=` that seeds it.
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go]: Correct, and it is the sharpest of the set. `scale := int64(1)` has no binary right-hand side, so the DEFINE branch could never match it — a comment claiming coverage it did not have, inside the test written to stop exactly that. Comment corrected to say what the branch does catch (`scale = scale * 10` written out) and to name why the seeding form cannot reach it.

File: backend/minorunitscale_test.go [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+			}
+		case *ast.SelectorExpr:
+			switch v.Sel.Name {
+			case "FormatInt", "Itoa", "Sprintf", "Sprint":
+				formats = true
+			}
+		}
+		return true
+	})
+	return point && formats
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go]: Correct. The digit-count parameter must now be REFERENCED in the body, not merely present in the signature. Planted as `a digit count the render ignores`.

File: backend/minorunitscale_test.go [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+// that looked for the asking.
+func rendersADecimalPoint(fn *ast.FuncDecl) bool {
+	if !takesADigitCount(fn) {
+		return false
+	}
+	point, formats := false, false
+	ast.Inspect(fn, func(n ast.Node) bool {
+		switch v := n.(type) {
+		case *ast.BasicLit:
+			if v.Kind == token.STRING && (v.Value == `"."` || strings.Contains(v.Value, `.%0`)) {
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go:140]: Addressed by the same change as the parameter thread: `point`, `formats` and a reference to the digit-count parameter must all hold. The residue that remains is real and stated rather than hidden — the three signals are collected over the whole body, so a function that renders a decimal for one reason and uses the digit count for another is still reachable. Narrowing further needs data flow; what the test does instead is name the boundary and plant both directions of it, so the next author knows the pairing is per-function and not per-expression.

File: backend/minorunitscale_test.go [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+	name := digitParamName(fn)
+	ast.Inspect(fn, func(n ast.Node) bool {
+		switch v := n.(type) {
+		case *ast.BasicLit:
+			if v.Kind == token.STRING && (v.Value == `"."` || strings.Contains(v.Value, `.%0`)) {
+				point = true
+			}
+		case *ast.Ident:
+			if v.Name == name {
+				uses = true
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go:145]: Fixed. `ast.Inspect(fn, …)` walks Type.Params, so the parameter's own identifier in the SIGNATURE satisfied the guard before the body was read — a guard that could not fail, in the file written to stop exactly that. It walks `fn.Body` now, and `a renderer whose digit count is signature-only` is planted for it. Mutation-verified: restoring the whole-declaration walk turns the suite red.

File: backend/minorunitscale_test.go [hunk 9]
```diff
@@ -0,0 +1,530 @@
...
+// built by iterating the digit count; a lone `x * 10` is a different thing
+// entirely, and counting it made the gate fire on any function that read the
+// digit count for a legitimate reason and happened to multiply something by
+// ten elsewhere. The detector still cannot prove the loop is bounded BY the
+// digit count — that needs data flow — but "in a loop" is the part that
+// distinguishes a scale from arithmetic, and it is checkable.
+func buildsAPowerOfTen(fn *ast.FuncDecl) bool {
+	builds := false
+	ast.Inspect(fn, func(n ast.Node) bool {
+		if sel, ok := n.(*ast.SelectorExpr); ok && sel.Sel.Name == "Pow10" {
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go:262]: Fixed. `Pow10` is resolved through the file's math import now, so an unrelated `legacy.Pow10` is not this defect. Mutation-verified.

File: backend/minorunitscale_test.go [hunk 10]
```diff
@@ -0,0 +1,530 @@
...
+func asksForDigits(fn *ast.FuncDecl, qualifier string) bool {
+	asks := false
+	ast.Inspect(fn, func(n ast.Node) bool {
+		call, ok := n.(*ast.CallExpr)
+		if !ok {
+			return true
+		}
+		switch f := call.Fun.(type) {
+		case *ast.Ident:
+			asks = asks || (f.Name == digitsFunc && qualifier == "")
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go:237]: Fixed. The census requires the file to BE the values package for a bare call to count, rather than inferring it from a missing import. Planted as `another package's own unqualified MinorUnitDigits`, and the probes gained a third rendering mode so the case can be written at all — the detector's answers depend on which file the code is in, so a probe that guesses wrong asks a different question than the tree does.

File: backend/minorunitscale_test.go [hunk 11]
```diff
@@ -0,0 +1,530 @@
...
+		case *ast.SelectorExpr:
+			pkg, ok := f.X.(*ast.Ident)
+			asks = asks || (ok && f.Sel.Name == digitsFunc && qualifier != "" && pkg.Name == qualifier)
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go:240]: Correct, and it is now stated rather than implied away. A dot-imported call is a bare identifier, indistinguishable from a local function of the same name without type information, so the census would rather miss it than name the wrong function. The doc comment says that, and nothing in this tree dot-imports values.

File: backend/minorunitscale_test.go [hunk 12]
```diff
@@ -0,0 +1,603 @@
...
+func handScaled(file *ast.File) []string {
+	var found []string
+	valuesQualifier := importedAs(file)
+	// A bare `MinorUnitDigits(...)` is values' own call only inside values. In
+	// any other package it names something else entirely, and reporting it
+	// sends somebody to disprove a finding.
+	if file.Name != nil && file.Name.Name == "values" {
+		valuesQualifier = ""
+	} else if valuesQualifier == "" {
+		return nil
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go]: Correct, and the sharpest finding of the round: the early exit disabled BOTH arms, and the renderer arm never asks values anything — being keyed on the parameter is the entire reason it can see the fourth copy. So the short-circuit re-blinded the census to exactly the shape the arm was added for, one commit after adding it. The values check gates the arithmetic arm only now. Planted as `a renderer in a file that does not import values`, verified to go red when the early exit is put back.

File: backend/minorunitscale_test.go [hunk 13]
```diff
@@ -0,0 +1,603 @@
...
+			return spec.Name.Name
+		}
+		return "values"
+	}
+	return ""
+}
+
+// mathAs returns the local name the math package is bound to, or "" if this
+// file does not import it.
+func mathAs(file *ast.File) string {
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/minorunitscale_test.go]: Correct, and pointed at the right PR. They are one `importAlias` now — and they had already diverged on the dot-import case by the time you flagged it, which is the argument this whole change is making, arriving inside it for the third time.
